spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb

```diff
@@ -2,154 +2,172 @@
+      expect(instance.class).to receive(:direct_otp_length).and_return(10)
+      instance.create_direct_otp
+      expect(instance.direct_otp.length).to equal(10)
 
-        it 'matches Devise configured length' do
-          expect(subject.length).to eq(Devise.otp_length)
-        end
+      expect(instance.class).to receive(:direct_otp_length).and_return(6)
+      instance.create_direct_otp
+      expect(instance.direct_otp.length).to equal(6)
+    end
 
-        context 'with a known time' do
-          let(:time) { 1_392_852_756 }
+    it "honors 'direct_otp_length' in options paramater" do
+      instance.create_direct_otp({length: 8})
+      expect(instance.direct_otp.length).to equal(8)
+      instance.create_direct_otp({length: 10})
+      expect(instance.direct_otp.length).to equal(10)
+    end
+  end
 
-          it 'returns a known result' do
-            expect(subject).
-              to eq('0000000524562202'.split(//).last(Devise.otp_length).join)
-          end
-        end
+  describe '#authenticate_direct_otp' do
+    let(:instance) { build_guest_user }
+    it 'fails if no direct_otp has been set' do
+      expect(instance.authenticate_direct_otp('12345')).to eq(false)
+    end
 
-        context 'with a known time yielding a result with less than 6 digits' do
-          let(:time) { 1_393_065_856 }
+    context 'after generating an OTP' do
+      before :each do
+        instance.create_direct_otp
+      end
 
-          it 'returns a known result padded with zeroes' do
-            expect(subject).
-              to eq('0000001608007672'.split(//).last(Devise.otp_length).join)
-          end
-        end
+      it 'accepts correct OTP' do
+        Timecop.freeze(Time.now + instance.class.direct_otp_valid_for - 1.second)
+        expect(instance.authenticate_direct_otp(instance.direct_otp)).to eq(true)
+      end
+
+      it 'rejects invalid OTP' do
+        Timecop.freeze(Time.now + instance.class.direct_otp_valid_for - 1.second)
+        expect(instance.authenticate_direct_otp('12340')).to eq(false)
+      end
+
+      it 'rejects expired OTP' do
+        Timecop.freeze(Time.now + instance.class.direct_otp_valid_for + 1.second)
+        expect(instance.authenticate_direct_otp(instance.direct_otp)).to eq(false)
       end
-    end
 
-    it_behaves_like 'otp_code', GuestUser.new
-    it_behaves_like 'otp_code', EncryptedUser.new
+      it 'revents code re-use' do
+        expect(instance.authenticate_direct_otp(instance.direct_otp)).to eq(true)
+        expect(instance.authenticate_direct_otp(instance.direct_otp)).to eq(false)
+      end
+    end
   end
 
-  describe '#authenticate_otp' do
-    shared_examples 'authenticate_otp' do |instance|
+  describe '#authenticate_totp' do
+
+    shared_examples 'authenticate_totp' do |instance|
       before :each do
         instance.otp_secret_key = '2z6hxkdwi3uvrnpn'
+        @totp_helper = TotpHelper.new(instance.otp_secret_key, instance.class.otp_length)
       end
```

Comment:
Before: We needed to know the current TOTP value at runtime so we could deliver it to the user.
Now: We don't need to know the TOTP values at runtime at all.. we only need to verify the ones that user enters.